src/map.rs

```diff
@@ -0,0 +1,114 @@
+use std::{marker::PhantomData, sync::Arc};
+
+use crate::Service;
+
+pub trait MapService<S1: Service, S2: Service>: std::fmt::Debug + Send + Sync + 'static {
+    fn req_up(&self, req: S2::Req) -> S1::Req;
+    fn res_up(&self, res: S2::Res) -> S1::Res;
+    fn try_req_down(&self, req: S1::Req) -> Result<S2::Req, ()>;
+    fn try_res_down(&self, res: S1::Res) -> Result<S2::Res, ()>;
+}
+
+#[derive(Debug)]
+pub struct Mapper<S1, S2> {
```

Comment:
So if you don't want to use the mapping feature at all this will be a dyn call for every request and every response message? Seems acceptable, there are no allocations.